docs/extension-e2e-test-guidelines.md

```diff
@@ -0,0 +1,528 @@
+| driver.delay(timeInMillis) | test/e2e/snaps/test-snap-\*\*\*.spec.js | 43 | Investigate excessive usage. Identify potential opportunities to implement a waiting strategy specific to snaps. |
+
+### Proposal
+
+Gradually remove the delay variables, starting with ones that are hardly used.
+| Function | Usage | Notes |
+| --- | --- | --- |
+| veryLargeDelayMs | 15 | Minimal usage in files other than metamask-ui.spec.js. Candidate for removal. Likely to be easy to remove once we complete the migration to fixtures. |
+| largeDelayMs | 30 | Minimal usage in files other than metamask-ui.spec.js. Candidate for removal. Likely to be easy to remove once we complete the migration to fixtures. |
+| regularDelayMs | 60 | Widespread usage |
+| tinyDelayMs | 14 | Mostly used in performance tests |
+
+### Repeatable tests
+
+Avoiding conditional statements in the end-to-end tests promotes simpler and more maintainable tests. A clear linear flow helps with understanding especially when it comes to identifying and fixing problems in failed tests. Changes in the extension's behaviour can easily break tests that rely on specific conditions to be met. It’s important to write repeatable tests that produce the same result every time they are run regardless of the environment. Conditional statements add variability to the tests as well as increase the test's complexity, making it harder to control the outcome. Changes to the test will need to be made in multiple places as the test logic is spread across branches.
+
+### Guidelines
+
+| Type             | Use | Notes                                                |
+| ---------------- | --- | ---------------------------------------------------- |
+| If statement     | ❌  | Introduces complexity and variability into the tests |
+| Ternary operator | ❌  | Introduces complexity and variability into the tests |
+| Switch           | ❌  | Introduces complexity and variability into the tests |
+| Short circuiting | ❌  | Introduces complexity and variability into the tests |
+
+### Proposal
+
+❌ Current condition code in test:
+
+```javascript
+if (type !== signatureRequestType.signTypedData)
+```
+
+Proposed solution: Remove conditional statements in test, structure the tests in a way that avoids branching logic.
+
+## Assertions <a name="assertions"></a>
+
+Assertions should verify expected behaviour and outcomes during test execution. When tests are responsible for assertions rather than being hidden in helper functions, it promotes test readability as expected results are explicitly stated. This makes it easier for engineers to quickly understand the intent of the test as well as the cause of failures.
+
+### Guidelines
+
+Please review the following video to get a better understanding of the guidelines below: [Assertions proposal.mov](https://drive.google.com/file/d/1ocwm0mwqXyBYgzuDB17OVYbQ2naHz4mR/view)
```

Comment:
Nice catch, thank you. This video explains the race condition, which is already explained in the document. I think I will just remove the link to avoid redundancy.